crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2026,6 +2027,80 @@ impl<'db> Class<'db> {
         self.iter_mro(db).contains(&ClassBase::Class(other))
     }
 
+    /// Return the explicit `metaclass` of this class, if one is defined.
+    fn explicit_metaclass(self, db: &'db dyn Db) -> Option<Type<'db>> {
+        let class_stmt = self.node(db);
+        let metaclass = class_stmt
+            .keywords()
+            .iter()
+            .find(|keyword| keyword.arg.as_ref().is_some_and(|arg| arg == "metaclass"))
+            .map(|keyword| &keyword.value);
+        if class_stmt.type_params.is_some() {
+            // when we have a specialized scope, we'll look up the inference
+            // within that scope
+            let model = SemanticModel::new(db, self.file(db));
+            metaclass.map(|base| base.ty(&model))
+        } else {
+            // Otherwise, we can do the lookup based on the definition scope
+            let class_definition = semantic_index(db, self.file(db)).definition(class_stmt);
+            metaclass.map(|base_node| definition_expression_ty(db, class_definition, base_node))
+        }
+    }
+
+    /// Return the metaclass of this class.
+    ///
+    /// ## Note
+    /// Only call this function from queries in the same file or your
+    /// query depends on the AST of another file (bad!).
```

Comment:
Calling this function from other files is fine, because you now added the `#[salsa::tracked`] which "isolates" the dependencies (dependent queries only re-run when the returned `Result` is different from previous run)
```suggestion
```